calyx/opt/src/passes_experimental/fsm_builder.rs

```diff
@@ -0,0 +1,276 @@
+                                        other_assigns.push(assign.clone())
+                                    })
+                                    .or_insert(vec![assign]);
+                            })
+                    });
+                    self.state += sen.group.borrow().latency;
+                    // Don't know where to transition next; let the parent that called
+                    // `build_abstract_fsm_with_loop` deal with registering the transition
+                    // from the state(s) we just built.
+                    (
+                        vec![IncompleteTransition::new(
+                            self.state - 1,
+                            ir::Guard::True,
+                        )],
+                        looped_once_guard,
+                    )
+                } else {
+                    // must be cyclic otherwise
+                    // placeholder code for now.
+                    self.register_transitions(
+                        self.state,
+                        &mut transitions_to_curr,
+                        guard.clone(),
+                    );
+
+                    let final_state_guard =
+                        self.leave_one_state_condition(guard, sen);
+
+                    self.state += 1;
+                    (
+                        vec![IncompleteTransition::new(
+                            self.state - 1,
+                            final_state_guard,
+                        )],
+                        None,
+                    )
+                }
+            }
+            ir::StaticControl::Seq(sseq) => (
+                sseq.stmts.iter().fold(
+                    transitions_to_curr,
+                    |transitions_to_this_stmt, stmt| {
+                        self.build_abstract(
+                            stmt,
+                            guard.clone(),
+                            transitions_to_this_stmt,
+                            looped_once_guard.clone(),
+                        )
+                        .0
+                    },
+                ),
+                None,
+            ),
+            ir::StaticControl::Repeat(_srep) => {
+                todo!()
+            }
+            ir::StaticControl::If(_sif) => {
+                todo!()
+            }
+            ir::StaticControl::Par(_spar) => {
+                todo!()
+            }
+            ir::StaticControl::Invoke(_) => {
+                unreachable!(
+                    "`build_abstract_cyclic` encountered a `static_invoke` node. \
+              Should have been compiled away."
+                )
+            }
+        }
+    }
+
+    fn fsm_build(
+        &mut self,
+        control: &ir::StaticControl,
+        build_kind: Option<bool>, // need to get better type name. Some(True) means non-promoted-static-component. False means promoted/static island. Otherwise it's a
+    ) -> ir::RRC<ir::FSM> {
+        let true_guard = ir::Guard::True;
```

Comment:
I'm not sure this name makes the code any more readable… this is only used once, and using `ir::Guard::True` might actually be clearer there.